File: Labs/experiment2/C_jump_sort.cpp
#include <iostream>
#include <cstdlib>
using namespace std;

int lst[100005];

int gcd(int x, int y) {
    if (x < y) swap(x, y);
    if (x % y == 0) return y;
    return gcd(y, x % y);
}

int main() {
    int n, k, g = 0;
    bool flag = false;
    cin >> n >> k;
    for (int i = 1; i <= n; i++) {
        cin >> lst[i];
        if (lst[i] - i == 0) continue;
        if (!flag) g = abs(lst[i] - i), flag = true;
        else g = gcd(g, abs(lst[i] - i));
    }
    while (k--) {
        int tmp;
        cin >> tmp;
        if (g % tmp == 0) cout << "YES" << endl;
        else cout << "NO" << endl;
    }
    return 0;
}

File: Labs/experiment2/C_jump_sort.py
def gcd(a, b):
    if a < b:
        a, b = b, a
    if a % b == 0:
        return b
    return gcd(b, a % b)

n, k = map(int, input().split())
lst = [0] + list(map(int, input().split()))
g = 0
flag = False
for i in range(1, n + 1):
    if lst[i] - i == 0:
        continue
    if not flag:
        g = abs(lst[i] - i)
        flag = True
    else:
        g = gcd(g, abs(lst[i] - i))

for _ in range(k):
    tmp = int(input())
    print("YES" if g % tmp == 0 else "NO")
